Editor/AddressableImporter.cs

```diff
@@ -26,47 +26,81 @@ static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAsse
-                        else
-                            Debug.LogFormat("[AddressableImporter] Entry created/updated for {0} with address {1}", assetPath, entry.address);
-                    }
-                }
+                var group = AssetDatabase.LoadAssetAtPath<AddressableAssetGroup>(importedAsset);
+                if (group == null || group.IsDefaultGroup()|| group.entries.Count >= 1 || !settings.groups.Contains(group))
+                    continue;
+                
+                settings.RemoveGroup(group);
+                dirty = true;
             }
         }
-        if (entriesAdded.Count > 0)
+        foreach (var movedAsset in movedAssets)
         {
-            settings.SetDirty(AddressableAssetSettings.ModificationEvent.EntryMoved, entriesAdded, true);
-            AssetDatabase.SaveAssets();
-        }
-        if (importSettings.removeEmtpyGroups)
-        {
-            settings.groups.RemoveAll(_ => _.entries.Count == 0 && !_.IsDefaultGroup());
+            if (prefabStage == null || prefabStage.prefabAssetPath != movedAsset) // Ignore current editing prefab asset.
+                dirty |= ApplyImportRule(movedAsset, settings, importSettings);
         }
+        
+        if (dirty)
+            AssetDatabase.SaveAssets();
     }
 
     static AddressableAssetGroup CreateAssetGroup<SchemaType>(AddressableAssetSettings settings, string groupName)
     {
         return settings.CreateGroup(groupName, false, false, false, new List<AddressableAssetGroupSchema> { settings.DefaultGroup.Schemas[0] }, typeof(SchemaType));
     }
 
+    static bool ApplyImportRule(
+        string assetPath, 
+        AddressableAssetSettings settings,
+        AddressableImportSettings importSettings)
+    {
+        var dirty = false;
+        var matched = false;
+        foreach (var rule in importSettings.rules)
+        {
+            if (!rule.Match(assetPath))
+                continue;
+
+            var entry = CreateOrUpdateAddressableAssetEntry(settings, importSettings, rule, assetPath);
+            if (entry != null)
+            {
+                if (rule.HasLabel)
+                    Debug.LogFormat("[AddressableImporter] Entry created/updated for {0} with address {1} and labels {2}", assetPath, entry.address, string.Join(", ", entry.labels));
+                else
+                    Debug.LogFormat("[AddressableImporter] Entry created/updated for {0} with address {1}", assetPath, entry.address);
+            }
+
+            dirty = true;
+            matched = true;
+            break;
+        }
+
+        if (!matched)
+        {
+            var guid = AssetDatabase.AssetPathToGUID(assetPath);
+            if (settings.RemoveAssetEntry(guid))
+            {
+                dirty = true;
+                Debug.LogFormat("[AddressableImporter] Entry removed for {0}", assetPath);
+            }
+        }
```

Comment:
We need more discussion here. The importer can not assume all assets are managed by it. This logic will destroy any manual imported addressable entry. Ideally, the remove logic should only apply to the asset that was managed by the importer before. But seems no way to tell, as we don't know where the asset moved from. One way I come with is by adding a label to mark an asset as managed, but adding extra data is not ideal.

Or at least we should add a new setting field "remove unmanaged entry", which explicitly warns the user of the side effect.

Reply:
You are right. The importer should not remove any unmanaged entries. It was my oversight.

I think "where the asset moved from" can get from the fourth argument of `OnPostprocessAllAssets`. So I will try to implement it after tomorrow.

Reply:
Fixed this at 77e799f.

Reply:
Looks good.